ACI - added secure environment variables

neilpeterson opened the pull request:
This PR adds Azure Container Instances [secure environment variables](https://docs.microsoft.com/en-us/azure/container-instances/container-instances-environment-variables#secure-values). 

Fundamentally the only difference between an environment variable and a secured variable is the value property name (Value vs. SecureValue), however for ease of use I am treating them as separate properties (seen in the below sample).

```
resource "azurerm_container_group" "both" {
  name                = "both"
  location            = "${azurerm_resource_group.vote-resource-group.location}"
  resource_group_name = "${azurerm_resource_group.vote-resource-group.name}"
  ip_address_type     = "public"
  os_type             = "linux"

  container {
    name   = "both"
    image  = "neilpeterson/nepetersv1"
    cpu    = "0.5"
    memory = "1.5"
    port   = "80"

    environment_variables {
      "NON_SECURE"  = "This is exposed in the portal and logs etc..."
    }

    secure_environment_variables {
      "SECURE"  = "This is not exposed in the portal and logs etc..."
    }

  }
}
```

neilpeterson commented:
@metacpp in the old PR, you ask to rev the ACI SDK version. It seems this version introduces some changes which will require a refactor of the Terraform provider. Do you want the refactor + my additional functionality added in one PR, or would it be ok for me to add the functionality first, and then follow up with a PR to refactor the provider to the new SDK version?

https://github.com/terraform-providers/terraform-provider-azurerm/pull/1874#discussion_r222491972

neilpeterson commented:
This one is done.

https://github.com/terraform-providers/terraform-provider-azurerm/pull/1874#discussion_r222495487

neilpeterson commented:
This one is done

https://github.com/terraform-providers/terraform-provider-azurerm/pull/1874#discussion_r222503395

neilpeterson commented:
This one is done.

https://github.com/terraform-providers/terraform-provider-azurerm/pull/1874#discussion_r222503426

neilpeterson commented:
@metacpp can you point me towards the docs so that I can update. I am unable to locate them.

neilpeterson commented:
Outstanding Issues from corrupt PR:

https://github.com/terraform-providers/terraform-provider-azurerm/pull/1874#discussion_r222494693

https://github.com/terraform-providers/terraform-provider-azurerm/pull/1874#discussion_r222503208

https://github.com/terraform-providers/terraform-provider-azurerm/pull/1874#discussion_r222503269

metacpp commented:
> @metacpp in the old PR, you ask to rev the ACI SDK version. It seems this version introduces some changes which will require a refactor of the Terraform provider. Do you want the refactor + my additional functionality added in one PR, or would it be ok for me to add the functionality first, and then follow up with a PR to refactor the provider to the new SDK version?
> 
> [#1874 (comment)](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1874#discussion_r222491972)

@neilpeterson please go ahead with current SDK version.

metacpp commented:
> @metacpp can you point me towards the docs so that I can update. I am unable to locate them.

Please add description for `secure_environment_variables` at:
https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/website/docs/r/container_group.html.markdown

metacpp commented on azurerm/resource_arm_container_group.go:
```diff
@@ -169,6 +169,13 @@ func resourceArmContainerGroup() *schema.Resource {
+						"secure_environment_variables": {
+							Type:      schema.TypeMap,
+							Optional:  true,
+							ForceNew:  true,
```
Shall we `ForceNew` here?

metacpp commented on azurerm/resource_arm_container_group.go:
```diff
@@ -599,18 +622,33 @@ func expandContainerGroupContainers(d *schema.ResourceData) (*[]containerinstanc
+func expandContainerEnvironmentVariables(input interface{}, secure bool) *[]containerinstance.EnvironmentVariable {
+
 	envVars := input.(map[string]interface{})
 	output := make([]containerinstance.EnvironmentVariable, 0)
```
output := make([]containerinstance.EnvironmentVariable, 0, len(envVars)) might be helpful to reduce reallocation. Depends on you, it might not be so helpful coz the size of variables will be limited.

metacpp commented on azurerm/resource_arm_container_group_test.go:
```diff
@@ -469,6 +475,12 @@ resource "azurerm_container_group" "test" {
+	secure_environment_variables {
+		"secureFoo"  = "secureBar"
+		"secureFoo1" = "secureBar1"
+	}
```
Can you replace tabs with whitespaces ? I know there're many inconsistency within this file, but at least we should be consistency for this particular configuration.

metacpp commented on vendor/vendor.json:
```diff
@@ -43,12 +43,10 @@
-			"revisionTime": "2018-09-28T00:20:07Z",
-			"version": "v21.1.0",
-			"versionExact": "v21.1.0"
+			"revisionTime": "2018-09-27T22:44:43Z"
```
You missed `version` here. Please use this command:
`govendor fetch github.com/Azure/azure-sdk-for-go/services/containerinstance/mgmt/2018-06-01/containerinstance@v21.1.0`

metacpp commented on vendor/vendor.json:
```diff
@@ -467,12 +465,16 @@
+			"path": "github.com/Azure/go-autorest/logger",
+			"revision": "a35eae345f69bbfbe3b8fa0b1d3fe98f8430b21a",
+			"revisionTime": "2018-08-30T19:44:05Z"
+		},
```
Do we need to update go-autorest/logger here?

metacpp commented on vendor/vendor.json:
```diff
@@ -467,12 +465,16 @@
+		},
+		{
+			"checksumSHA1": "scpSozMdk4sqSpkbQqupLKUfLiM=",
 			"path": "github.com/Azure/go-autorest/version",
```
Do we need to change it?

neilpeterson replied:
I was unable to successfully run `make vendor-status` without this change.

metacpp commented on vendor/vendor.json:
```diff
@@ -1509,6 +1511,12 @@
+			"path": "golang.org/x/text/internal/tag",
+			"revision": "6e3c4e7365ddcc329f090f96e4348398f6310088",
+			"revisionTime": "2018-08-07T15:59:08Z"
+		},
```
Where do we use this package?

neilpeterson commented:
@metacpp - thanks, I've addressed all comments. I am unsure on this one, changes seem necessary. 

https://github.com/terraform-providers/terraform-provider-azurerm/pull/2024#discussion_r224283548

I noticed a few additional changes to `vendor.json` that were not related to my updates. I've validated in each case that `make vendor-status` fails without the changes. 

Finally, not sure how this crept in, but `make vet` detected and unreachable return in `resource_arm_api_management.go`. The test looks valid to me and I've removed the return.

https://github.com/terraform-providers/terraform-provider-azurerm/pull/2024/commits/331f71a126390a7c46c5499c3112b145182455db

Thanks

metacpp commented:
@neilpeterson Did some fix on the test failures in https://github.com/terraform-providers/terraform-provider-azurerm/pull/2024/commits/d779d67adeec49f948af079b55572bb10bd0f8ec

neilpeterson commented:
@metacpp thanks a bunch. I had checked the build and everything looked good. My apologies for these issues.

metacpp commented on azurerm/resource_arm_container_group.go:
```diff
@@ -34,11 +34,11 @@ func resourceArmContainerGroup() *schema.Resource {
 				Type:             schema.TypeString,
 				Optional:         true,
-				Default:          "Public",
+				Default:          "public",
```
Please assign it with API spec: https://github.com/Azure/azure-rest-api-specs/blob/master/specification/containerinstance/resource-manager/Microsoft.ContainerInstance/stable/2018-06-01/containerInstance.json#L1054

metacpp commented on azurerm/resource_arm_container_group.go:
```diff
@@ -34,11 +34,11 @@ func resourceArmContainerGroup() *schema.Resource {
 				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
 				ValidateFunc: validation.StringInSlice([]string{
-					"Public",
+					"public",
```
Align with API spec.

metacpp commented on azurerm/resource_arm_container_group.go:
```diff
@@ -166,21 +166,24 @@ func resourceArmContainerGroup() *schema.Resource {
-							ForceNew: true,
+						},
+
+						"sensitive_environment_variables": {
```
Can we align with Azure documentation ?
https://docs.microsoft.com/en-us/azure/container-instances/container-instances-environment-variables#secure-values

jeffreyCline replied:
I disagree, as this implies a higher level of security that isn't there since the values are persisted in the HCL file plan test. Since it is not encrypted I would feel better if we would expose it as sensitive and in the code pass it as secureValue as required by the API.

paultyng replied:
I think there was a misunderstanding that this wasn't surfacing something the API here internally. While these variable are not "secure" in Terraform, they are more secure in the container group and that is the name used in the API, so we should probably match it.

jeffreyCline replied:
Agreed, I have reverted back to using the secure name instead of sensitive.

jeffreyCline commented:
![image](https://user-images.githubusercontent.com/20408400/47129086-b91d8d00-d248-11e8-8a91-76df8659db45.png)

katbyte commented:
tests pass:
<img width="499" alt="screen shot 2018-10-18 at 00 06 24" src="https://user-images.githubusercontent.com/1638467/47136855-cf881080-d269-11e8-9849-a72b7d52de31.png">
